File: src/main/java/com/back/domain/search/search/service/SearchService.java
package com.back.domain.search.search.service;

import com.back.domain.auction.auction.entity.Auction;
import com.back.domain.auction.auction.repository.AuctionRepository;
import com.back.domain.post.post.entity.Post;
import com.back.domain.post.post.repository.PostRepository;
import com.back.domain.search.search.dto.UnifiedSearchResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class SearchService {
    private final PostRepository postRepository;
    private final AuctionRepository auctionRepository;

    // 통합 검색 (POST + AUCTION)
    public Page<UnifiedSearchResponse> searchUnified(String keyword, Pageable pageable) {
        // 1. Post 검색
        Page<Post> posts = postRepository.search(keyword, pageable);

        // 2. Auction 검색
        Page<Auction> auctions = auctionRepository.search(keyword, pageable);

        // 3. 결과를 UnifiedSearchResponse로 변환
        List<UnifiedSearchResponse> combinedResults = new ArrayList<>();

        // Post 변환
        posts.forEach(post -> combinedResults.add(
            UnifiedSearchResponse.builder()
                .id(post.getId())
                .type("POST")
                .title(post.getTitle())
                .price(post.getPrice())
                .status(post.getStatus().name())
                .categoryName(post.getCategory().getName())
                .thumbnailUrl(post.getPostImages().isEmpty() ? null
                    : post.getPostImages().get(0).getImage().getUrl())
                .createDate(post.getCreateDate())
                .build()
        ));

        // Auction 변환
        auctions.forEach(auction -> combinedResults.add(
            UnifiedSearchResponse.builder()
                .id(auction.getId())
                .type("AUCTION")
                .title(auction.getName())
                .price(auction.getStartPrice())
                .status(auction.getStatus().name())
                .categoryName(auction.getCategory().getName())
                .thumbnailUrl(auction.getAuctionImages().isEmpty() ? null
                    : auction.getAuctionImages().get(0).getImage().getUrl())
                .createDate(auction.getCreateDate())
                .build()
        ));

        // 4. 최신순 정렬
        List<UnifiedSearchResponse> sortedResults = combinedResults.stream()
            .sorted(Comparator.comparing(UnifiedSearchResponse::getCreateDate).reversed())
            .collect(Collectors.toList());

        // 5. 페이징 처리
        int start = (int) pageable.getOffset();
        int end = Math.min((start + pageable.getPageSize()), sortedResults.size());

        List<UnifiedSearchResponse> pagedResults = start >= sortedResults.size()
            ? new ArrayList<>()
            : sortedResults.subList(start, end);

        return new PageImpl<>(pagedResults, pageable, sortedResults.size());
    }
}

File: src/main/java/com/back/domain/search/search/service/SearchService.kt
package com.back.domain.search.search.service

import com.back.domain.search.search.dto.SearchCursor
import com.back.domain.search.search.dto.SearchResponse
import com.back.domain.search.search.dto.UnifiedSearchResponse
import com.back.domain.search.search.service.port.UnifiedSearchRepository
import com.back.domain.search.search.service.projection.CursorCodec
import com.back.domain.search.search.service.projection.UnifiedSearchRow
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Slice
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class SearchService(
    /*
        private val postRepository: PostRepository,
        private val auctionRepository: AuctionRepository

     */
    //private val searchProviders: List<SearchProvider>
    private val unifiedSearchRepository: UnifiedSearchRepository
) {

    // 통합 검색 (POST + AUCTION)
    fun searchUnified(keyword: String, sort: String, pageable: Pageable): Slice<UnifiedSearchResponse> {
        val kw = toBooleanKw(keyword)
        val p = PageRequest.of(pageable.pageNumber, pageable.pageSize) // Sort 제거

        val rows = when (sort.lowercase()) {
            "newest" -> unifiedSearchRepository.searchUnifiedNewest(kw, p)
            "oldest" -> unifiedSearchRepository.searchUnifiedOldest(kw, p)
            else -> unifiedSearchRepository.searchUnifiedRelevance(kw, p) // default relevance
        }

        return rows.map { it.toDto() }
    }

    fun searchUnifiedCursor(keyword: String, sort: String, size: Int, cursor: String?): SearchResponse {
        val kw = toBooleanKw(keyword)
        val pageable = PageRequest.of(0, size.coerceIn(1, 50))

        val cur = cursor?.let { CursorCodec.decode(it) }

        val slice = when (sort.lowercase()) {
            "newest" -> unifiedSearchRepository.searchNewestCursor(
                kw,
                cur?.createDate, cur?.typeRank, cur?.id,
                pageable
            )
            "oldest" -> unifiedSearchRepository.searchOldestCursor(
                kw,
                cur?.createDate, cur?.typeRank, cur?.id,
                pageable
            )
            else -> unifiedSearchRepository.searchRelevanceCursor(
                kw,
                cur?.score, cur?.createDate, cur?.typeRank, cur?.id,
                pageable
            )
        }

        val content = slice.content.map { it.toDto() } // score/typeRank 버림

        val nextCursor = slice.content.lastOrNull()?.let { last ->
            val nc = SearchCursor(
                sort = sort.lowercase(),
                score = last.score,               // newest/oldest는 null
                createDate = last.createDate,
                typeRank = last.typeRank,
                id = last.id
            )
            CursorCodec.encode(nc)
        }

        return SearchResponse.ofCursor(
            content = content,
            size = pageable.pageSize,
            hasNext = slice.hasNext(),
            nextCursor = nextCursor
        )
    }

    private fun normalizeSort(sort: String): String =
        when (sort.lowercase()) {
            "relevance", "newest", "oldest" -> sort.lowercase()
            else -> "relevance"
        }

    fun toBooleanKw(input: String): String =
        input.trim()
            .split(Regex("\\s+"))
            .filter { it.isNotBlank() }
            .joinToString(" ") { "+${it}*" }

    private fun UnifiedSearchRow.toDto() = UnifiedSearchResponse(
        id = id,
        type = type,
        title = title,
        price = price,
        status = status,
        statusDisplayName = statusDisplayName,
        categoryId = categoryId,
        thumbnailUrl = thumbnailUrl,
        createDate = createDate
    )

//    fun searchUnifiedCursor(
//        keyword: String,
//        sort: String,
//        size: Int,
//        cursor: String?
//    ): SearchResponse {
//        return when (sort) {
//            "newest" -> searchNewest(keyword, size, cursor)
//            "oldest" -> searchOldest(keyword, size, cursor)
//            else -> searchRelevance(keyword, size, cursor)
//        }
//    }
//
//    private fun searchRelevance(
//        keyword: String,
//        size: Int,
//        cursor: String?
//    ): SearchResponse {
//        val decoded = RelevanceCursor.decode(cursor)
//
//        val slice = searchDocumentMvQueryService.searchRelevance(
//            keyword = keyword,
//            cursorScore = decoded?.score,
//            cursorCreateDate = decoded?.createDate,
//            cursorId = decoded?.id,
//            size = size
//        )
//
//        val items = slice.content.map { it.toResponse() }
//
//        val nextCursor = if (slice.hasNext() && items.isNotEmpty()) {
//            val last = items.last()
//            RelevanceCursor(
//                score = last.score ?: 0.0,
//                createDate = last.createDate,
//                id = last.id
//            ).encode()
//        } else {
//            null
//        }
//
//        return SearchResponse(
//            content = items,
//            page = slice.number,
//            size = slice.size,
//            hasNext = slice.hasNext(),
//            nextCursor = nextCursor
//        )
//    }
//
//    private fun searchNewest(
//        keyword: String,
//        size: Int,
//        cursor: String?
//    ): SearchResponse {
//        val decoded = DateCursor.decode(cursor)
//
//        val slice = searchDocumentMvQueryService.searchNewest(
//            keyword = keyword,
//            cursorCreateDate = decoded?.createDate,
//            cursorId = decoded?.id,
//            size = size
//        )
//
//        val items = slice.content.map { it.toResponse() }
//
//        val nextCursor = if (slice.hasNext() && items.isNotEmpty()) {
//            val last = items.last()
//            DateCursor(
//                createDate = last.createDate,
//                id = last.id
//            ).encode()
//        } else {
//            null
//        }
//
//        return SearchResponse(
//            content = items,
//            page = slice.number,
//            size = slice.size,
//            hasNext = slice.hasNext(),
//            nextCursor = nextCursor
//        )
//    }
//
//    private fun searchOldest(
//        keyword: String,
//        size: Int,
//        cursor: String?
//    ): SearchResponse {
//        val decoded = DateCursor.decode(cursor)
//
//        val slice = searchDocumentMvQueryService.searchOldest(
//            keyword = keyword,
//            cursorCreateDate = decoded?.createDate,
//            cursorId = decoded?.id,
//            size = size
//        )
//
//        val items = slice.content.map { it.toResponse() }
//
//        val nextCursor = if (slice.hasNext() && items.isNotEmpty()) {
//            val last = items.last()
//            DateCursor(
//                createDate = last.createDate,
//                id = last.id
//            ).encode()
//        } else {
//            null
//        }
//
//        return SearchResponse(
//            content = items,
//            page = slice.number,
//            size = slice.size,
//            hasNext = slice.hasNext(),
//            nextCursor = nextCursor
//        )
//    }
//
//    private fun UnifiedSearchRow.toResponse(): UnifiedSearchResponse {
//        return UnifiedSearchResponse(
//            id = id,
//            type = type,
//            title = title,
//            price = price,
//            status = status,
//            statusDisplayName = statusDisplayName,
//            categoryId = categoryId,
//            thumbnailUrl = thumbnailUrl,
//            createDate = createDate,
//            score = score
//        )
//    }
}